File: backend/src/mhglauncher/services/wishes.py
from __future__ import annotations

import asyncio
import builtins
from collections import defaultdict
from collections.abc import Callable
from typing import TYPE_CHECKING

from mhglauncher.database import Database
from mhglauncher.models import GameRole, WishBannerDetail, WishRecord, WishStatistics
from mhglauncher.providers.base import Provider
from mhglauncher.services.item_metadata import enrich_record, remote_icon_urls
from mhglauncher.services.wish_statistics import build_banner_detail

if TYPE_CHECKING:
    from mhglauncher.services.image_cache import ImageCacheService

WishLog = Callable[[str], None]


class WishService:
    def __init__(
        self,
        database: Database,
        provider: Provider,
        image_cache: ImageCacheService | None = None,
        port: int = 0,
    ) -> None:
        self.database = database
        self.provider = provider
        self._image_cache = image_cache
        self._port = port

    async def sync(
        self,
        credential: str,
        role: GameRole,
        log: WishLog | None = None,
    ) -> int:
        inserted = 0
        newest_ids = await self._newest_ids(role.uid)
        if log is not None:
            log(f"已读取 {len(newest_ids)} 个卡池的本地增量检查点")
        page_count = 0
        async for page in self.provider.iter_wishes(credential, role, newest_ids):
            before = await self._count(role.uid)
            await self.save(page)
            added = await self._count(role.uid) - before
            inserted += added
            page_count += 1
            if log is not None:
                log(f"第 {page_count} 页读取 {len(page)} 条记录，新增 {added} 条")
        if log is not None:
            log(f"米游社分页读取完成，共处理 {page_count} 页")
        return inserted

    async def _newest_ids(self, uid: str) -> dict[str, str]:
        rows = await self.database.fetch_all(
            """
            SELECT
              COALESCE(NULLIF(uigf_gacha_type, ''), gacha_type) AS gacha_type,
              MAX(id) AS id
            FROM wishes WHERE uid=?
            GROUP BY COALESCE(NULLIF(uigf_gacha_type, ''), gacha_type)
            """,
            (uid,),
        )
        return {str(row["gacha_type"]): str(row["id"]) for row in rows}

    async def save(self, records: list[WishRecord]) -> None:
        values = [
            (
                item.id,
                item.uid,
                item.gacha_type,
                item.uigf_gacha_type,
                item.item_id,
                item.name,
                item.item_type,
                item.rank,
                item.time.isoformat(),
            )
            for item in records
        ]
        if not values:
            return
        async with self.database.connect() as connection:
            await connection.executemany(
                """
                INSERT INTO wishes(
                  id, uid, gacha_type, uigf_gacha_type,
                  item_id, name, item_type, rank, time
                ) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(id) DO UPDATE SET
                  uigf_gacha_type=COALESCE(
                    NULLIF(excluded.uigf_gacha_type, ''),
                    wishes.uigf_gacha_type
                  ),
                  name=COALESCE(NULLIF(excluded.name, ''), wishes.name),
                  item_type=COALESCE(NULLIF(excluded.item_type, ''), wishes.item_type),
                  rank=CASE WHEN excluded.rank > 0 THEN excluded.rank ELSE wishes.rank END
                """,
                values,
            )

    async def clear(self) -> int:
        async with self.database.connect() as connection:
            cursor = await connection.execute("DELETE FROM wishes")
            return max(cursor.rowcount, 0)

    async def list(self, uid: str, gacha_type: str | None = None) -> list[WishRecord]:
        sql = "SELECT * FROM wishes WHERE uid=?"
        values: list[str] = [uid]
        if gacha_type:
            sql += " AND gacha_type=?"
            values.append(gacha_type)
        sql += " ORDER BY time DESC, id DESC"
        rows = await self.database.fetch_all(sql, values)
        records = []
        item_ids: set[str] = set()
        for row in rows:
            data = dict(row)
            data["uigf_gacha_type"] = data["uigf_gacha_type"] or _uigf_type(str(data["gacha_type"]))
            record = WishRecord.model_validate(data)
            item_ids.add(record.item_id)
            records.append(enrich_record(record, self._image_cache, self._port))
        if item_ids and self._image_cache is not None:
            urls = remote_icon_urls(item_ids)
            if urls:
                asyncio.ensure_future(self._image_cache.ensure_all(urls))  # noqa: RUF006
        return records

    async def statistics(self, uid: str) -> builtins.list[WishStatistics]:
        records = await self.list(uid)
        groups: dict[str, builtins.list[WishRecord]] = defaultdict(builtins.list)
        for record in records:
            groups[record.uigf_gacha_type].append(record)
        results = []
        for gacha_type, items in sorted(groups.items()):
            pulls = 0
            for item in items:
                if item.rank == 5:
                    break
                pulls += 1
            results.append(
                WishStatistics(
                    uid=uid,
                    gacha_type=gacha_type,
                    total=len(items),
                    five_star_count=sum(item.rank == 5 for item in items),
                    pulls_since_five_star=pulls,
                )
            )
        return results

    async def _count(self, uid: str) -> int:
        row = await self.database.fetch_one(
            "SELECT COUNT(*) AS count FROM wishes WHERE uid=?",
            (uid,),
        )
        return int(row["count"]) if row else 0

    async def banner_statistics(self, uid: str) -> builtins.list[WishBannerDetail]:
        sql = "SELECT * FROM wishes WHERE uid=? ORDER BY time ASC, id ASC"
        rows = await self.database.fetch_all(sql, (uid,))
        records = []
        for row in rows:
            data = dict(row)
            data["uigf_gacha_type"] = data["uigf_gacha_type"] or _uigf_type(str(data["gacha_type"]))
            records.append(WishRecord.model_validate(data))

        groups: dict[str, builtins.list[WishRecord]] = defaultdict(builtins.list)
        for record in records:
            groups[record.uigf_gacha_type].append(record)

        results = []
        for gacha_type, items in sorted(groups.items()):
            results.append(
                build_banner_detail(uid, gacha_type, items, self._image_cache, self._port)
            )
        return results


def _uigf_type(gacha_type: str) -> str:
    return "301" if gacha_type == "400" else gacha_type

File: backend/src/services/wishes.ts
import type { Store } from "../core/database";
import type { CompanionSnapshot, GameRole, WishRecord, WishStatistics } from "../core/models";
import type { Provider } from "../providers/provider";
import type { GameRecordSource } from "../providers/game-record";
import type { ImageCache } from "./images";
import { enrich } from "./metadata";
import { banner, type BannerDetail } from "./wish-statistics";
import { AppError } from "../core/errors";

const wish = (row: Record<string, unknown>): WishRecord => ({
  id: String(row.id), uid: String(row.uid), gacha_type: String(row.gacha_type),
  uigf_gacha_type: String(row.uigf_gacha_type || (row.gacha_type === "400" ? "301" : row.gacha_type)),
  item_id: String(row.item_id), name: String(row.name), item_type: String(row.item_type),
  rank: Number(row.rank), time: String(row.time),
});

export class WishService {
  constructor(
    private readonly store: Store, private readonly provider: Provider,
    private readonly images: ImageCache, private readonly gachaRecords?: GameRecordSource,
  ) {}

  async sync(credential: string, role: GameRole, log?: (message: string) => void): Promise<number> {
    const newest = Object.fromEntries(this.store.all(`SELECT gacha_type,id FROM (
      SELECT COALESCE(NULLIF(uigf_gacha_type,''),gacha_type) gacha_type,id,
      ROW_NUMBER() OVER(PARTITION BY COALESCE(NULLIF(uigf_gacha_type,''),gacha_type) ORDER BY LENGTH(id) DESC,id DESC) row_number
      FROM wishes WHERE uid=?) WHERE row_number=1`, role.uid)
      .map((row) => [String(row.gacha_type), String(row.id)]));
    log?.(`已读取 ${Object.keys(newest).length} 个卡池的本地增量检查点`);
    let inserted = 0, pages = 0;
    for await (const records of this.provider.wishes(credential, role, newest)) {
      const added = this.newRecordCount(records); this.save(records);
      inserted += added; pages += 1; log?.(`第 ${pages} 页读取 ${records.length} 条记录，新增 ${added} 条`);
    }
    log?.(`米游社分页读取完成，共处理 ${pages} 页`);
    return inserted;
  }

  async importFromGachaUrl(url: string, log?: (message: string) => void): Promise<{ inserted: number; uids: string[] }> {
    if (!this.gachaRecords) throw new AppError("gacha_import_unavailable", "当前环境不支持抽卡 URL 导入", 503);
    let inserted = 0, pages = 0;
    const uids = new Set<string>();
    for await (const records of this.gachaRecords.wishesFromGachaUrl(url)) {
      records.forEach(({ uid }) => uids.add(uid));
      if (uids.size > 1) throw new AppError("gacha_uid_mismatch", "抽卡 URL 返回了不一致的 UID", 422);
      const added = this.newRecordCount(records); this.save(records);
      inserted += added; pages += 1; log?.(`第 ${pages} 个卡池读取 ${records.length} 条记录，新增 ${added} 条`);
    }
    log?.(`抽卡 URL 分页读取完成，共处理 ${pages} 个卡池`);
    return { inserted, uids: [...uids] };
  }

  save(records: WishRecord[]): void {
    if (!records.length) return;
    const insert = this.store.db.prepare(`INSERT INTO wishes(id,uid,gacha_type,uigf_gacha_type,item_id,name,item_type,rank,time,time_epoch)
      VALUES(?,?,?,?,?,?,?,?,?,?) ON CONFLICT(uid,id) DO UPDATE SET
      uigf_gacha_type=COALESCE(NULLIF(excluded.uigf_gacha_type,''),wishes.uigf_gacha_type),
      name=COALESCE(NULLIF(excluded.name,''),wishes.name),item_type=COALESCE(NULLIF(excluded.item_type,''),wishes.item_type),
      rank=CASE WHEN excluded.rank>0 THEN excluded.rank ELSE wishes.rank END`);
    this.store.db.transaction(() => records.forEach((value) => {
      const time = normalizeTime(value.time);
      if (!time) throw new AppError("wish_time_invalid", "祈愿时间格式无效", 422);
      insert.run(value.id, value.uid, value.gacha_type, value.uigf_gacha_type, value.item_id,
        value.name, value.item_type, value.rank, time.iso, time.epoch);
    }))();
  }

  clear(): number { return this.store.db.prepare("DELETE FROM wishes").run().changes; }

  list(uid: string, type?: string): WishRecord[] {
    const rows = type
      ? this.store.all("SELECT * FROM wishes WHERE uid=? AND gacha_type=? ORDER BY time_epoch DESC,LENGTH(id) DESC,id DESC", uid, type)
      : this.store.all("SELECT * FROM wishes WHERE uid=? ORDER BY time_epoch DESC,LENGTH(id) DESC,id DESC", uid);
    return rows.map(wish).map((value) => enrich(value, this.images));
  }

  statistics(uid: string): WishStatistics[] {
    const rows = this.store.all(`SELECT uid,COALESCE(NULLIF(uigf_gacha_type,''),gacha_type) uigf_gacha_type,rank
      FROM wishes WHERE uid=? ORDER BY time_epoch DESC,LENGTH(id) DESC,id DESC`, uid);
    return this.statisticsFrom(rows.map((row) => ({
      id: "", uid: String(row.uid), gacha_type: String(row.uigf_gacha_type),
      uigf_gacha_type: String(row.uigf_gacha_type), item_id: "", name: "", item_type: "",
      rank: Number(row.rank), time: "",
    })));
  }

  snapshot(uid: string, note: unknown): CompanionSnapshot {
    const wishes = this.list(uid), ascending = [...wishes].reverse();
    return {
      wishes, statistics: this.statisticsFrom(wishes),
      banner_statistics: [...Map.groupBy(ascending, (value) => value.uigf_gacha_type).entries()]
        .sort().map(([type, values]) => banner(uid, type, values, this.images)),
      note: note as CompanionSnapshot["note"],
    };
  }

  private statisticsFrom(records: WishRecord[]): WishStatistics[] {
    const groups = Map.groupBy(records, (value) => value.uigf_gacha_type);
    return [...groups.entries()].sort().map(([type, records]) => ({
      uid: records[0]?.uid ?? "", gacha_type: type, total: records.length,
      five_star_count: records.filter(({ rank }) => rank === 5).length,
      pulls_since_five_star: records.findIndex(({ rank }) => rank === 5) < 0 ? records.length : records.findIndex(({ rank }) => rank === 5),
    }));
  }

  bannerStatistics(uid: string): BannerDetail[] {
    const records = this.store.all("SELECT * FROM wishes WHERE uid=? ORDER BY time_epoch ASC,LENGTH(id) ASC,id ASC", uid).map(wish);
    return [...Map.groupBy(records, (value) => value.uigf_gacha_type).entries()].sort()
      .map(([type, values]) => banner(uid, type, values, this.images));
  }

  private newRecordCount(records: WishRecord[]): number {
    const ids = [...new Set(records.map(({ id }) => id))];
    if (!ids.length) return 0;
    const placeholders = ids.map(() => "?").join(",");
    const uid = records[0]?.uid ?? "";
    const existing = new Set(this.store.all(`SELECT id FROM wishes WHERE uid=? AND id IN (${placeholders})`, uid, ...ids).map(({ id }) => String(id)));
    return ids.filter((id) => !existing.has(id)).length;
  }
}

function normalizeTime(value: string): { iso: string; epoch: number } | null {
  const explicit = /(Z|[+-]\d{2}:\d{2})$/i.test(value) ? value : `${value.replace(" ", "T")}+08:00`;
  const epoch = Date.parse(explicit);
  return Number.isFinite(epoch) ? { iso: new Date(epoch).toISOString(), epoch } : null;
}
